velox/functions/prestosql/aggregates/tests/GeometricMeanTest.cpp

```diff
@@ -0,0 +1,200 @@
+      makeFlatVector<int64_t>(
+          100,
+          [](auto row) { return row; },
+          [](auto row) {
+            // All values in group 3 are null.
+            if (row / 10 == 3) {
+              return true;
+            }
+
+            // Every other value is null.
+            return row % 2 == 0;
+          }),
+  });
+
+  auto expected = makeRowVector({
+      makeFlatVector<int32_t>({0, 1, 2, 3, 4, 5, 6, 7, 8, 9}),
+      makeFlatVector<double>(
+          10,
+          [](auto row) {
+            double logSum = 0;
+            int64_t count = 0;
+            for (int32_t i = 1; i < 10; i += 2) {
+              logSum += std::log(row * 10 + i);
+              count++;
+            }
+            return std::exp(logSum / count);
+          },
+          [](auto row) { return row == 3; }),
+  });
+
+  testAggregations({data}, {"c0"}, {"geometric_mean(c1)"}, {expected});
+}
+
+TEST_F(GeometricMeanTest, groupByIntegers) {
+  auto data = makeRowVector({
+      makeFlatVector<int32_t>(100, [](auto row) { return row / 10; }),
+      makeFlatVector<int64_t>(100, [](auto row) { return row; }),
+  });
+
+  auto expected = makeRowVector({
+      makeFlatVector<int32_t>({0, 1, 2, 3, 4, 5, 6, 7, 8, 9}),
+      makeFlatVector<double>(
+          10,
+          [](auto row) {
+            double logSum = 0;
+            int64_t count = 0;
+            for (int32_t i = 0; i < 10; ++i) {
+              logSum += std::log(row * 10 + i);
+              count++;
+            }
+            return std::exp(logSum / count);
+          }),
+  });
+
+  testAggregations({data}, {"c0"}, {"geometric_mean(c1)"}, {expected});
+}
+
+TEST_F(GeometricMeanTest, globalDoubles) {
+  auto data = makeRowVector({
+      makeFlatVector<double>(100, [](auto row) { return row * 0.1 / 7; }),
+  });
+
+  double logSum = 0;
+  int64_t count = 0;
+  for (int32_t i = 0; i < 100; ++i) {
+    logSum += std::log(i * 0.1 / 7);
+    count++;
+  }
+  double geometricMean = std::exp(logSum / count);
+
+  auto expected = makeRowVector({
+      makeFlatVector(std::vector<double>{
+          geometricMean,
+      }),
+  });
+
+  testAggregations({data}, {}, {"geometric_mean(c0)"}, {expected});
+}
+
+TEST_F(GeometricMeanTest, groupByDoubles) {
+  auto data = makeRowVector({
+      makeFlatVector<int32_t>(100, [](auto row) { return row / 10; }),
+      makeFlatVector<double>(100, [](auto row) { return row * 0.1; }),
+  });
+
+  auto expected = makeRowVector({
+      makeFlatVector<int32_t>({0, 1, 2, 3, 4, 5, 6, 7, 8, 9}),
+      makeFlatVector<double>(
+          10,
+          [](auto row) {
+            double logSum = 0;
+            int64_t count = 0;
+            for (int32_t i = 0; i < 10; ++i) {
+              logSum += std::log(row + i * 0.1);
+              count++;
+            }
+            return std::exp(logSum / count);
+          }),
+  });
+
+  testAggregations({data}, {"c0"}, {"geometric_mean(c1)"}, {expected});
+}
+
```

Comment:
Would it make sense to also add a test case that compares the aggregation result with DuckDB geometric_mean? See examples in velox/functions/prestosql/aggregates/tests/ArrayAggTest.cpp.

Reply:
The version of DuckDB seems does not have `geometric_mean` yet:

```
Function:verifyDuckDBResult, Expression: result->success DuckDB query failed: Catalog Error: Scalar Function with name geometric_mean does not exist!
Did you mean "median"?
LINE 1: SELECT geometric_mean(c0) FROM tmp
               ^
SELECT geometric_mean(c0) FROM tmp, Source: RUNTIME, ErrorCode: INVALID_STATE
```

And I checked with DuckDB's source code, seems geometric_mean is added recently: https://github.com/duckdb/duckdb/blame/239f51293c429168774c3943e96ddf2451253a07/src/catalog/default/default_functions.cpp#L102

Reply:
@kagamiori do I need to upgrade DuckDB to add the test?

Reply:
Upgrading is not possible. There is a GitHub issue about that with details. Look it up. We are working on changing AggregationFuzzer to use Presto as source of truth: https://github.com/facebookincubator/velox/issues/6595 Once this is done we'll be able to verify all aggregate functions.